Studies/source/GenStudy.cxx

```diff
@@ -0,0 +1,591 @@
+
+            const auto gen_jets = CreateGenJets(baryons_plus_mesons, 0.4);
+
+            //control for one jet cases
+            std::set<size_t> one_jet;
+            if(gen_jets.size() == 1)
+                one_jet.insert(1);
+
+            anaData.n("jets").Fill(gen_jets.size());
+
+            if(gen_jets.size() < 2) continue;
+
+            for(size_t jet_index = 0; jet_index < 2; ++jet_index){
+                for(size_t particle_index = 0; particle_index < gen_jets.at(jet_index).size(); ++particle_index){
+                    HH_Gen_Event.b_jets[jet_index] += gen_jets.at(jet_index).at(particle_index)->momentum;
+                }
+            }
+
+            for(size_t jet_index = 0; jet_index < gen_jets.size(); ++jet_index){
+                for(size_t particle_index = 0; particle_index < gen_jets.at(jet_index).size(); ++particle_index){
+                    HH_Gen_Event.b_jets_all[jet_index] += gen_jets.at(jet_index).at(particle_index)->momentum;
+                }
+            }
+            if(isTauInsideAcceptance(HH_Gen_Event, args.channel()))
+                anaData.n("pass_taus_acceptance").Fill(isTauInsideAcceptance(HH_Gen_Event, args.channel()));
+            if(isBInsideAcceptance(HH_Gen_Event))
+            anaData.n("pass_b_acceptance").Fill(isBInsideAcceptance(HH_Gen_Event));
+
+
+            if(!isTauInsideAcceptance(HH_Gen_Event, args.channel())) continue;
+            if(!isBInsideAcceptance(HH_Gen_Event)) continue;
+
+            anaData.n("pass_b_plus_taus_acceptance").Fill(isBInsideAcceptance(HH_Gen_Event));
+
+            HH_Gen_Event.h_bb_vis = (HH_Gen_Event.b_jets[0] + HH_Gen_Event.b_jets[1]);
+
+            for(size_t jet_index = 2; jet_index < gen_jets.size(); ++jet_index){
+                HH_Gen_Event.h_bb_vis_all += HH_Gen_Event.b_jets_all[jet_index];
+            }
+
+            //Matching Taus
+            std::set<size_t> tau_matches_total;
+            std::map<size_t, std::set<size_t>> tau_matches;
+
+            for (size_t tau_index = 0; tau_index < HH_Gen_Event.tau.size(); tau_index++) {
+                for (size_t reco_tau_index = 0; reco_tau_index < event.lep_p4.size(); reco_tau_index++) {
+                    if(args.channel() != Channel::TauTau && reco_tau_index != 0 &&
+                        HasMatchWithMCObject(event.lep_p4.at(0), event.lep_p4.at(reco_tau_index), 0.2)) continue;
+                    const auto& visibleMomentum = HH_Gen_Event.vis_tau[tau_index];
+                    HH_Gen_Event.h_tautau_vis += HH_Gen_Event.vis_tau[tau_index];
+                    auto dEta = HH_Gen_Event.tau[tau_index]->momentum.Eta() - event.lep_p4.at(reco_tau_index).Eta();
+                    auto dPhi = HH_Gen_Event.tau[tau_index]->momentum.Phi() - event.lep_p4.at(reco_tau_index).Phi();
+                    anaData.deta_tau().Fill(dEta);
+                    anaData.dphi_tau().Fill(dPhi);
+                    anaData.delta_eta_vs_delta_phi().Fill(dEta, dPhi);
```

Comment:
create separate deta_tau, dphi_tau and delta_eta_vs_delta_phi for different tau types